### ISSUE TITLE
Use the Boyer-Moore search algorithm for String.IndexOf

### ISSUE THREAD

**Issue by jamesqo** [dotnet/coreclr#6918]
The current implementation of `String.IndexOf(string)` [uses a naive loop](https://github.com/dotnet/coreclr/blob/master/src/utilcode/newapis.cpp#L598-L608) to check whether a substring of some given text matches a pattern. Although this implementation is simple and easy to understand, it is very inefficient because it has to iterate through a minimum of *m - n + 1* characters of the text, where *m* is the length of the text and *n* is the length of the pattern string.

We should instead use the [Boyer-Moore search algorithm](https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_string_search_algorithm) to implement this function. It performs much better for larger patterns, since it allows us to 'skip over' some characters based on what we know from pre-processing the string. [Here](http://stackoverflow.com/questions/6207819/boyer-moore-algorithm-understanding-and-example) is an answer on StackOverflow that explains how it works, and the Wikipedia link has sample implementations in C/Java.

**Comment by jkotas** [comment]
>  it is very inefficient

It is inefficient for large strings. It is pretty efficient for small strings that is likely the typical usage pattern for IndexOf...

**Comment by benaadams** [comment]
For smaller strings, could check first, a middle and last char before checking the intermediate chars (pre-processing will have some setup, so will be cross over point)

e.g. for words the entropy increases with each char; however there are also a lot of shared endings `-ing` `-ed` `-s` etc

**Comment by benaadams** [comment]
Also would only apply to non-culture strings as culture aware strings use OS compare (at least on windows)

**Comment by jamesqo** [comment]
@jkotas Yeah, for small strings it may be problematic since it needs a dynamically allocated buffer for `delta2`. For those scenarios I'm planning to use a 'lite' Boyer-Moore and use a tiny statically allocated bitmap, like what's being done with [IndexOfAny](https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/String.cs#L2127) right now. If any of the 2 bits corresponding to a char are not set in the bitmap, then we will be 100% sure it's not in the pattern and will still be able to apply the 'skip `pattern.Length` chars if that char isn't in the pattern' optimization. That IMO is the main benefit of this algorithm.

**Comment by jasonwilliams200OK** [comment]
@SoftCircuits, saw your varied implementations of Boyer-Moore algorithm via http://stackoverflow.com/a/4916363. Would love to have your take on this. :)

**Comment by jamesqo** [comment]
@benaadams 

> Also would only apply to non-culture strings as culture aware strings use OS compare (at least on windows)

I wonder if we could bypass doing that if `IsAscii()` is true, since I'm pretty sure any sane OS would return uniform results for those strings.

**Comment by SoftCircuits** [comment]
In my experience, `String.IndexOf()` with `StringComparision.Ordinal` will always be faster than Boyer-Moore except for extreme pathological cases. The reason appears to be that `String.IndexOf()` uses hand-optimized assembly language specifically designed to quickly scan strings. In addition, with Unicode, Boyer-Moore require a _very_ large buffer to be allocated or for the skip table unless you do a couple of tricks, which could potentially make it even slower. For comparisons other than `StringComparison.Ordinal`, a faster version might be possible. My article [Fast Text Search with Boyer-Moore](http://www.blackbeltcoder.com/Articles/algorithms/fast-text-search-with-boyer-moore) presents a C# implementation of Boyer-Moore and discusses all the issues mentioned here.

**Comment by jamesqo** [comment]
@SoftCircuits 

> hand-optimized assembly language instructions specifically designed to quickly scan strings

Where did you get that? From what [this SO answer](http://stackoverflow.com/a/2584204/4077294) has told me, the `IndexOf` implementation is written [here](https://github.com/dotnet/coreclr/blob/master/src/utilcode/newapis.cpp#L598-L608) in C++, not assembly.

> with Unicode, Boyer-Moore require a very large buffer to be allocated for the skip table unless you perform a couple of tricks

Yeah, I realized that earlier. There are a couple of workarounds for this. For example, String has a [cached `IsAscii` flag](https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/String.cs#L3906), so if we only run this algorithm on ASCII strings we will only have to allocate a `int[]` of size 128 (which can be done on the stack).

**Comment by benaadams** [comment]
Is will be hard to beat 2 strings in cpu cache by bringing in a 3rd data structure. Like to work better for file size searches and/or much larger search strings.

**Comment by jamesqo** [comment]
@benaadams Hm, maybe. I could see it being less efficient with the bitmap scheme because of all of the bit operations that have to be done (couple of shl/shr/and). However, if we go down the lookup table route that will only take up like 8 cache lines, which is only 1/64 of L1 cache so it won't cause problems with eviction. A lookup table will also introduce no additional branching to the main loop (see implementation [here](https://gist.github.com/jamesqo/9a34e9edc4a9452a021653ff357a5a24#file-boyermoore-cs-L164)), and it is O(n) to initialize depending on the length of the string so for smaller strings the added overhead may not be that much.

**Comment by benaadams** [comment]
But if you are looking for a < 32 char string that's only 1 cache line?

**Comment by SoftCircuits** [comment]
@jamesqo As I mentioned originally, the code used depends on the comparison type. I don't know what the code you referenced is (and don't have time to dig into it right now). What I do know is that the performance of `IndexOf()` (in some comparison modes) far surpassed anything I could write in C# (even Boyer-Moore algorithms) _at the time I ran my tests_. There were other clues at the time that indicated the code was unmanaged, and I have been told by the original .NET developers that a lot of the lower level routines were written in hand-optimized assembler. But I don't have first-hand knowledge of the exact code. And perhaps things have changed since I looked into it.
